checker/tests/resourceleak-arrays/owningArrayField.java

```diff
@@ -0,0 +1,230 @@
+    }
+  }
+
+  @EnsuresCalledMethodsOnElements(
+      value = "arr",
+      methods = {"foo"})
+  public void destruct() {
+    for (int i = 0; i < arr.length; i++) {
+      arr[i].foo();
+    }
+  }
+}
+
+class EvilOwningArrayWrapperClient {
+  public EvilOwningArrayWrapperClient() {
+    int n = 10;
+    @OwningArray Resource[] localarr = new Resource[n];
+    for (int i = 0; i < n; i++) {
+      localarr[i] = new Resource();
+    }
+    // give up ownership to constructor
+    ValidOwningArrayField d = new ValidOwningArrayField(localarr);
+    // this loop doesn't pattern-match, simply because d.arr is not accepted
+    // as an identifier (it is a MemberSelectTree). A warning is issued to inform
+    // the programmer that d.arr is an unexpected array expression.
+    for (int i = 0; i < d.arr.length; i++) {
+      // :: error: required.method.not.called
+      // :: error: illegal.owningarray.field.elements.assignment
+      // :: warning: unexpected.array.expression
+      d.arr[i] = new Resource();
+    }
+    // fulfill the obligations of 'd'
+    d.destruct();
+    d.close();
+  }
+
+  public void tryCapturingOwningArrayField() {
+    int n = 10;
+    @OwningArray Resource[] localarr = new Resource[n];
+    for (int i = 0; i < n; i++) {
+      localarr[i] = new Resource();
+    }
+    // give up ownership to constructor
+    ValidOwningArrayField d = new ValidOwningArrayField(localarr);
+    // try reassigning the elements of array, despite lost ownership
+    for (int i = 0; i < n; i++) {
+      // :: error: assignment.without.ownership
+      localarr[i] = new Resource();
+    }
+    // this method call is not allowed either, due to the missing ownership over localarr
+    // :: error: argument
+    // :: error: argument.with.revoked.ownership
+    methodWithOwningArrayParameter(localarr);
+
+    // reassign localarr to a new array, which is legal.
+    // However, after its elements have been assigned, they
+    // are never closed, which throws an error.
+    // :: error: unfulfilled.mustcallonelements.obligations
+    localarr = new Resource[n];
+    // assigning its elements is now allowed
+    for (int i = 0; i < n; i++) {
+      localarr[i] = new Resource();
+    }
+
+    // try to capture the @OwningArray field of 'd'
+    // since the RHS of the assignment is @OwningArray, that is illegal aliasing.
+    // :: error: illegal.aliasing
+    Resource[] capture = d.arr;
+    // this illegal assignment counts as a field assignment
+    // and since the field is @OwningArray, it is forbidden
+    // :: error: owningarray.field.assigned.outside.constructor
+    d.arr[0] = null;
+
+    // fulfill the obligations of 'd'
+    d.destruct();
+    d.close();
+  }
+
+  private void methodWithOwningArrayParameter(@OwningArray Resource[] arr) {
+    for (int i = 0; i < arr.length; i++) {
+      arr[i].close();
+    }
+    for (int i = 0; i < arr.length; i++) {
+      arr[i].foo();
+    }
+  }
+}
+
+class OwningArrayFieldClient {
```

Comment:
Can you add a copy of this test where `d.close()` does actually get called, to make sure that we have an example of a client that actually has no errors in this test case? I suspect that that might be useful to users, who will sometimes look at our test cases as examples of what to do/not do.